connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,266 @@
+muxer is found, the muxer selection process MUST fall back to multistream
+-selection protocol.
+
+Example: Noise handshake between peers that have a mutually supported muxer.
+    Responder supports: ["yamux/1.0.0", "/mplex/6.7.0"]
+    Initiator supports: ["yamux/1.0.0"]
+
+    XX:
+    -> e
+    <- e, ee, s, es, ["yamux/1.0.0", "/mplex/6.7.0"] 
+    -> s, se, ["yamux/1.0.0"]
+
+    After handshake is done, both parties can arrive on the same conclusion
+    and select "yamux/1.0.0" as the muxer to use.
+
+Example: Noise handshake between peers that don't have mutually supported
+muxers.
+    Responder supports: ["/mplex/6.7.0"]
+    Initiator supports: ["yamux/1.0.0"]
+
+    XX:
+    -> e
+    <- e, ee, s, es, ["/mplex/6.7.0"]
+    -> s, se, ["yamux/1.0.0"]
+    
+    After handshaking is done, early data processing will find no mutually
+    supported muxer, and falls back to multistream-selection protocol.
+
+The muxer selection logic is run outside of the Noise handshake process. The
+format of he early data for this purpose is specified in the protobuf in the
+[Early-data-specification] section.
+
+#### Early data specification
+
+The early data message is encoded in the "protobuf2" syntax as shown in the
+following. The protobuf definition is an extension to [handshake-payload]. The
+existing byte array early data (the "data" field) will be replaced by a
+structured NoiseExtensions protobuf message. The supported muxers and selected
+muxer are populated in the "stream_muxers" field. The details of the early
+data message can be find in [Noise-handshake-payload]
+
+The muxers are ordered by preference, with the most prefered muxer at the
+beginning.
+
+```protobuf
+syntax = "proto2";
+
+message NoiseExtensions {
+    repeated bytes webtransport_certhashes = 1;
+    repeated string stream_muxers = 2; 
+}
+
+message NoiseHandshakePayload {
+  optional bytes identity_key = 1;
+  optional bytes identity_sig = 2;
+  optional NoiseExtensions extensions = 4;
+}
+```
+
+## Cross version support
+
+The improved muxer selection approach MUST be interoperable with pervious
+libp2p versions which do not support this improved approach.
+
+### TLS case
+
+In the current version of libp2p, the ALPN extension field is populated with a
+key "libp2p". By appending the key of "libp2p" to the end of the supported
+muxer list, the TLS handshaking process is not broken when peers run different
+versions of libp2p, because the minimum overlap of the peer's supported muxer
+sets is always satisfied. When one peer runs the old version and the other peer
+runs the version that supports this feature, the negotiated protocol is
+"libp2p".
+
+In the case "libp2p" is the result of TLS ALPN, an empty result MUST be
+returned to the upgrade process to indicate that no muxer was selected. And the
+upgrade process MUST fall back to the multistream-selection protocol to
+to negotiate the muxer to be selected.
+
+### Noise case
```

Comment:
Explain how backwards compatibility works in the Noise section of this document?

Reply:
done